### PR TITLE
Calendar operators

### PR DESCRIPTION
**Pull request by ianspektor**
### Calendar operators

Only implemented `day_of_month` operator for now, which yields the day of month of each timestamp. Called it `calendar_day_of_month`, I propose prefixing all calendar operators with `calendar_` for easier understanding. 

I would also create a submodule in the operators folders for them, but I recall in some conversation we decided to have all operators in the top level module. @achoum lmk if you belive this still holds, since calendar operators will be 10 new operators in that module. I might also need to create a base class for some of these since logic will be very similar (converting floats to datetimes, then grabbing a property from the datetime).

Once we validate this one I'll move on to implementing all the ones described in the design doc:
```
weekday(), week_of_month(), week_of_year(), month(), year(), hours(), minutes(), seconds()
```

### Other unrelated changes

- Made all existing operators available in the top level API.
- Fixed some docstrings and class names in test modules (e.g. renamed AssignOperatorTest to AssignNumpyImplementationTest since it's testing the logic of the numpy impl)
- Fixed some type hints in __call__ return types

### CODE REVIEW

File: temporian/implementation/numpy/operators/calendar_day.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class CalendarDayNumpyImplementation:
+    def __init__(self, operator: CalendarDayOperator) -> None:
+        super().__init__()
+        self.operator = operator
+
+    def __call__(self, event: NumpyEvent) -> Dict[str, NumpyEvent]:
+        data = {}
+        for index, timestamps in event.sampling.data.items():
+            print(timestamps)
```
Comment (diegomarvid) [inline temporian/implementation/numpy/operators/calendar_day.py:32]: remove prints

File: temporian/core/operators/calendar_day.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    """
+
+    def __init__(self, event: Event):
+        super().__init__()
+
+        # input
+        self.add_input("event", event)
+
+        output_feature = Feature(
+            name="calendar_day",
```
Comment (diegomarvid) [inline temporian/core/operators/calendar_day.py:37]: I think calendar day its too generic. Could it be day of the week, day of the month, day of the year. Shouldn't be better to name it with something like `day_of_the_month` or `calendar_day_of_the_month`?

If we have **day of the week** with this naming convention it would be `calendar_week` and it's misleading if it's the day of the week (1-7) or the week of the year (1-52). In that case, I would prefer as well `day_of_the_week` or `calendar_day_of_the_week` instead of `calendar_week`.

**Review by diegomarvid — CHANGES_REQUESTED** [review]

File: temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+)
+from temporian.implementation.numpy.data.event import NumpyEvent
+from temporian.implementation.numpy.data.event import NumpyFeature
+from temporian.implementation.numpy.data.sampling import NumpySampling
+from temporian.implementation.numpy.operators.calendar_day_of_month import (
+    CalendarDayOfMonthNumpyImplementation,
+)
+
+
+class CalendarDayNumpyImplementationTest(absltest.TestCase):
```
Comment (diegomarvid) [inline temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py:32]: `CalendarDayOfMonthNumpyImplementationTest` too long but it is what it is 😆

Comment (ianspektor) [reply temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py:32]: missed it, changing it, nice catch

File: temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    def test_day(self) -> None:
+        """Test calendar day operator."""
+        input_sampling_data = NumpySampling(
+            index=[],
+            data={
+                (): np.array(
+                    [
+                        0,  # 1970-01-01
+                        1678895040,  # 2023-03-15
```
Comment (achoum) [inline temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py:43]: Assuming the figure is  UTC epoch, in Kiritimati, this is the 16th :)

**Review by achoum — APPROVED** [review]
This is great

File: temporian/implementation/numpy/operators/calendar_day_of_month.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+class CalendarDayOfMonthNumpyImplementation:
+    def __init__(self, operator: CalendarDayOfMonthOperator) -> None:
+        super().__init__()
+        self.operator = operator
+
+    def __call__(self, event: NumpyEvent) -> Dict[str, NumpyEvent]:
+        data = {}
+        for index, timestamps in event.sampling.data.items():
+            days = np.array(
+                [datetime.fromtimestamp(ts).day for ts in timestamps]
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/implementation/numpy/operators/calendar_day_of_month.py:35]: Currently, If the user sends a Date Object we convert the timestamp to a UTC epoch, so the timezone for every timestamp in the sampling will be UTC.

[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/implementation/numpy/operators/calendar_day_of_month.py:35]: Nice catch! Was planning on adding some edge case tests (e.g. at 00:01 and 23:59) to ensure exactly this, will do so and fix it. Agree with adding the timezone as an optional arg.

File: temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    """Test numpy implementation of calendar_day_of_month operator."""
+
+    def test_day(self) -> None:
+        """Test calendar day operator."""
+        input_sampling_data = NumpySampling(
+            index=[],
+            data={
+                (): np.array(
+                    [
+                        0,  # 1970-01-01
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply temporian/implementation/numpy/operators/tests/calendar_day_of_month_test.py:42]:
```python
date_1 = datetime.datetime.strptime("2021-01-01 23:00:00", "%Y-%m-%d %H:%M:%S")
date_1 = date_1.replace(
    tzinfo=datetime.timezone.tzname("Uruguay Standard Time")
)

event = NumpyEvent.from_dataframe(
    pd.DataFrame(
        data=[
            [date_1, 14],
        ],
        columns=["timestamp", "sales"],
    ),
)
```

**Review by diegomarvid — APPROVED** [review]

File: temporian/implementation/numpy/operators/calendar_day_of_month.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/implementation/numpy/operators/calendar_day_of_month.py:35]: from the link Mathieu posted:
> If tz is not None, it must be an instance of a [tzinfo](https://docs.python.org/3/library/datetime.html#datetime.tzinfo) subclass

skipping timezone argument for now since we only support numeric/string arguments for operators. creating issue off of this conversation for the future.

[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/implementation/numpy/operators/calendar_day_of_month.py:35]: added this to next meeting's notes 💯
